server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -313,12 +313,15 @@ public boolean isForceExecution() {
         });
     }
 
+
     @Override
     public final void executeNextPhase(SearchPhase currentPhase, SearchPhase nextPhase) {
         /* This is the main search phase transition where we move to the next phase. At this point we check if there is
          * at least one successful operation left and if so we move to the next phase. If not we immediately fail the
-         * search phase as "all shards failed"*/
-        if (successfulOps.get() == 0) { // we have 0 successful results that means we shortcut stuff and return a failure
+         * search phase as "all shards failed". Note that if we cancelled the rest of the search request because of a
+         * partial failure, we don't count this as if "all shards failed".
+         */
+        if (successfulOps.get() == 0 && requestCancelled.get() == false) {
```

Comment:
Drive by comment:

I wonder if we need to check if (shardFailures.size() == total) and if so, still report all shards failed? At least it will be a little bit confusing to get the partial shards failure exception if you only search one shard.

Reply:
I pushed an update to only check `shardSearchFailures.length == getNumShards()`. But I think this only makes sense with #64337 -- I'll pull that it once it's merged and check if there are still test failures.